pkg/controller/nodelifecycle/node_lifecycle_controller.go

```diff
@@ -488,20 +467,35 @@ func (nc *Controller) doFixDeprecatedTaintKeyPass(node *v1.Node) error {
 	return nil
 }
 
-func (nc *Controller) doNoScheduleTaintingPassWorker(i int, stopCh <-chan struct{}) {
+func (nc *Controller) doNoScheduleTaintingPassWorker() {
 	for {
-		select {
-		case <-stopCh:
+		obj, shutdown := nc.nodeUpdateQueue.Get()
+		// "nodeUpdateQueue" will be shutdown when "stopCh" closed;
+		// we do not need to re-check "stopCh" again.
+		if shutdown {
 			return
-		case node := <-nc.nodeUpdateChannels[i]:
-			if err := nc.doNoScheduleTaintingPass(node); err != nil {
-				glog.Errorf("Failed to taint NoSchedule on node <%s>: %v", node.Name, err)
-			}
 		}
+		nodeName := obj.(string)
+
+		if err := nc.doNoScheduleTaintingPass(nodeName); err != nil {
+			// Add nodeName back to the queue, it'll be requeue after `Done`.
+			nc.nodeUpdateQueue.Add(nodeName)
```

Comment:
If you want to requeue, this needs to use the ratelimiting workqueue and track number of retries per item so we don't hotloop on a persistently failing item. 

See use of NewRateLimitingQueue(DefaultControllerRateLimiter()) elsewhere, along with AddRateLimited()/Forget() for handling requeuing items. 

I think the switch to requeue on error can be done as a follow up, since nodes are still updated every few seconds in this release

Reply:
OK, removed .

Reply:
+1 to @liggitt - this is non-trivial, that's why i suggested a TODO (and solve it in follow-up).